controller/proxy-injector/webhook_test.go

```diff
@@ -142,9 +166,32 @@ func TestGetPatch(t *testing.T) {
 
 func getFakeReq(b []byte) *admissionv1beta1.AdmissionRequest {
 	return &admissionv1beta1.AdmissionRequest{
-		Kind:      metav1.GroupVersionKind{Kind: "Deployment"},
+		Kind:      metav1.GroupVersionKind{Kind: "Pod"},
 		Name:      "foobar",
 		Namespace: "linkerd",
 		Object:    runtime.RawExtension{Raw: b},
 	}
 }
+
+func ownerRetriever(p *v1.Pod) (string, string) {
+	return pkgK8s.Deployment, "owner-deployment"
+}
+
+// Added annotations are backed by map, so we need to sort before
+// being able to do the comparison
+func sortedPatch(patchJSON []byte) (unmarshalledPatch, error) {
```

Comment:
I realize that maps are a good representation for labels and annotations, but something about non-deterministic ordering of patch operations feels like not quite the right fit.

What do you think about backing labels and annotations with lists in the patch operation, and then inserting them in sorted order? That way the patch output is deterministic.

I took a crack at this. It seems to work, but I don't fully understand the implication of changing `patchOp.Value` from `map[string]string{}` to `[]string{}`:

```diff
diff --git a/controller/proxy-injector/webhook_test.go b/controller/proxy-injector/webhook_test.go
index 24b2d820..89568311 100644
--- a/controller/proxy-injector/webhook_test.go
+++ b/controller/proxy-injector/webhook_test.go
@@ -4,7 +4,6 @@ import (
 	"encoding/json"
 	"fmt"
 	"reflect"
-	"sort"
 	"testing"
 
 	"github.com/linkerd/linkerd2/controller/gen/config"
@@ -97,7 +96,7 @@ func TestGetPatch(t *testing.T) {
 		if err != nil {
 			t.Fatalf("Unexpected error: %s", err)
 		}
-		expectedPatch, err := sortedPatch(expectedPatchBytes)
+		expectedPatch, err := unmarshalPatch(expectedPatchBytes)
 		if err != nil {
 			t.Fatalf("Unexpected error: %s", err)
 		}
@@ -132,7 +131,7 @@ func TestGetPatch(t *testing.T) {
 					t.Fatalf("Did not expect injection for file '%s'", testCase.filename)
 				}
 
-				actualPatch, err := sortedPatch(patchJSON)
+				actualPatch, err := unmarshalPatch(patchJSON)
 				if err != nil {
 					t.Fatalf("Unexpected error: %s", err)
 				}
@@ -177,21 +176,12 @@ func ownerRetriever(p *v1.Pod) (string, string) {
 	return pkgK8s.Deployment, "owner-deployment"
 }
 
-// Added annotations are backed by map, so we need to sort before
-// being able to do the comparison
-func sortedPatch(patchJSON []byte) (unmarshalledPatch, error) {
+func unmarshalPatch(patchJSON []byte) (unmarshalledPatch, error) {
 	var actualPatch unmarshalledPatch
 	err := json.Unmarshal(patchJSON, &actualPatch)
 	if err != nil {
 		return nil, err
 	}
 
-	sort.Sort(actualPatch)
 	return actualPatch, nil
 }
-
-func (u unmarshalledPatch) Len() int      { return len(u) }
-func (u unmarshalledPatch) Swap(i, j int) { u[i], u[j] = u[j], u[i] }
-func (u unmarshalledPatch) Less(i, j int) bool {
-	return u[i]["path"].(string) < u[j]["path"].(string)
-}
diff --git a/pkg/inject/inject.go b/pkg/inject/inject.go
index 8f95ff18..6e0dab99 100644
--- a/pkg/inject/inject.go
+++ b/pkg/inject/inject.go
@@ -3,6 +3,7 @@ package inject
 import (
 	"encoding/json"
 	"fmt"
+	"sort"
 	"strconv"
 	"strings"
 
@@ -617,17 +618,17 @@ func (conf *ResourceConfig) injectObjectMeta(patch *Patch) {
 		if len(conf.pod.meta.Labels) == 0 {
 			patch.addPodLabelsRoot()
 		}
-		for k, v := range conf.pod.labels {
-			patch.addPodLabel(k, v)
+		for _, k := range sortedKeys(conf.pod.labels) {
+			patch.addPodLabel(k, conf.pod.labels[k])
 		}
 	}
 
-	for k, v := range conf.pod.annotations {
-		patch.addPodAnnotation(k, v)
+	for _, k := range sortedKeys(conf.pod.annotations) {
+		patch.addPodAnnotation(k, conf.pod.annotations[k])
 
 		// append any additional pod annotations to the pod's meta.
 		// for e.g., annotations that were converted from CLI inject options.
-		conf.pod.meta.Annotations[k] = v
+		conf.pod.meta.Annotations[k] = conf.pod.annotations[k]
 	}
 }
 
@@ -926,3 +927,14 @@ func (conf *ResourceConfig) proxyOutboundSkipPorts() string {
 	}
 	return strings.Join(ports, ",")
 }
+
+func sortedKeys(m map[string]string) []string {
+	keys := []string{}
+	for k, _ := range m {
+		keys = append(keys, k)
+	}
+
+	sort.Strings(keys)
+
+	return keys
+}
diff --git a/pkg/inject/patch.go b/pkg/inject/patch.go
index edba6e9c..d6ccc1ec 100644
--- a/pkg/inject/patch.go
+++ b/pkg/inject/patch.go
@@ -100,7 +100,7 @@ func (p *Patch) addPodLabelsRoot() {
 	p.patchOps = append(p.patchOps, &patchOp{
 		Op:    "add",
 		Path:  p.patchPathPodLabels,
-		Value: map[string]string{},
+		Value: []string{},
 	})
 }
 
@@ -117,7 +117,7 @@ func (p *Patch) addPodAnnotationsRoot() {
 	p.patchOps = append(p.patchOps, &patchOp{
 		Op:    "add",
 		Path:  p.patchPathPodAnnotations,
-		Value: map[string]string{},
+		Value: []string{},
 	})
 }
 ```

Let me know what you think.

Reply:
@siggy fully agreed :+1: 
I've applied your changes except for `patch.go`'s. `Value` is actually supposed to be a json object, so chaging it to `[]string{}` brakes the patching, so I discarded that change, and it all works :clap: